### PR TITLE
Expose graphql hocs to allow choosing them explicitly and improve typing

### LINKED ISSUE #2568: Typescript graphql() hoc issue with option.prop and ChildDataProps typings
**Intended outcome:**

The documentation TS examples should work.

**Actual outcome:**

The documentation TS examples do not compile due to typings related to the "prop" option and ChildDataProp

https://www.apollographql.com/docs/react/recipes/static-typing.html#props

**How to reproduce the issue:**

Create a project with the example here: https://www.apollographql.com/docs/react/recipes/static-typing.html#props

**Version**

- apollo-client@latest
- react-apollo@latest

### PR DESCRIPTION
**Pull request by danilobuerger**
Fixes #2568

* Allows user to explicitly choose the hoc they want to use instead of relying on the parsed document type.
* Enabled typescript user to profit from tighter typings by not using Partials for the child props.

### CODE REVIEW

**Comment by RIP21** [comment]
@helfer @hwillson anyone can take a look at this one?

**Comment by RIP21** [comment]
@benjamn @steelbrain @helfer @hwillson  anyone once again? :) 
We need it, It's simple as hell, but will be amazing to have for TS.
Also, it will require a bit of a change to docs eventually. But right now it's a mess to work with TS and HOCs. 
Please, please, please.

**Comment by RIP21** [comment]
@benjamn @steelbrain @helfer @hwillson @helfer 
I really sorry for being an annoying ass. But this is SO needed.

File: src/mutation-hoc.tsx [hunk 1]
```diff
@@ -12,11 +12,11 @@ import {
   GraphQLBase,
 } from './hoc-utils';
 
-export function mutation<
+export function withMutation<
```
Comment (rosskevin) [inline src/mutation-hoc.tsx]: FYI - This is a breaking change and relies on a semver bump.

Comment (rosskevin) [reply src/mutation-hoc.tsx]: My bad, we use rollup and this wasn't included before.

Comment (danilobuerger) [reply src/mutation-hoc.tsx]: Yeah, this has never been exported by the actual react-apollo module. Also was never documented as public api.

**Comment by rosskevin** [comment]
Thanks @danilobuerger

**Pull request merged**